### PR TITLE
Check signing threshold and set EPoS status only at last epoch

### PR DESCRIPTION
**Pull request by rlan35**
Do ComputeAndMutateEPOSStatus at the last epoch, instead of at every block.

In EPoS election candidacy, validators in current committee who doesn't sign enough blocks will be kicked out no matter what. For those who signed enough, the candidacy is determined by the Status. 

Tested locally

### CODE REVIEW

File: staking/availability/measure.go [hunk 1]
```diff
@@ -71,25 +69,8 @@ func BlockSigners(
 // BallotResult returns
 // (parentCommittee.Slots, payable, missing, err)
 func BallotResult(
-	bc engine.ChainReader, header *block.Header, shardID uint32,
+	parentHeader, header *block.Header, parentShardState *shard.State, shardID uint32,
 ) (shard.SlotList, shard.SlotList, shard.SlotList, error) {
-	// TODO ek – retrieving by parent number (blockNum - 1) doesn't work,
-	//  while it is okay with hash.  Sounds like DB inconsistency.
-	//  Figure out why.
-	parentHeader := bc.GetHeaderByHash(header.ParentHash())
```
Comment (rlan35) [inline staking/availability/measure.go]: these are moved out to break import cycle.

File: staking/availability/measure.go [hunk 2]
```diff
@@ -193,68 +166,63 @@ func ComputeCurrentSigning(
 	signed, toSign :=
 		new(big.Int).Sub(statsNow.NumBlocksSigned, snapSigned),
 		new(big.Int).Sub(statsNow.NumBlocksToSign, snapToSign)
-	leftToGo := blocksPerEpoch - toSign.Uint64()
 
 	computed := staking.NewComputed(
-		signed, toSign, leftToGo, numeric.ZeroDec(), true,
+		signed, toSign, 0, numeric.ZeroDec(), true,
```
Comment (fxfactorial) [inline staking/availability/measure.go]: hmm, why take it out, people wanted it, will affect RPC

File: node/node_handler.go [hunk 3]
```diff
@@ -494,10 +494,11 @@ func (node *Node) PostConsensusProcessing(
...
 					return
 				}
-				computed, err :=
-					availability.ComputeCurrentSigning(
-						snapshot, wrapper, shard.Schedule.BlocksPerEpoch(),
-					)
+				computed := availability.ComputeCurrentSigning(
+					snapshot, wrapper,
+				)
+				computed.BlocksLeftInEpoch = shard.Schedule.BlocksPerEpoch() - computed.ToSign.Uint64()
```
Comment (fxfactorial) [inline node/node_handler.go]: ah I see, you set the blocks left her

File: staking/availability/measure.go (hunk 2 above)
Comment (rlan35) [reply staking/availability/measure.go]: i moved it outside the func. it's only for RPCs. for protocol calculation, it's not needed.

File: staking/availability/measure.go (hunk 1 above)
Comment (fxfactorial) [reply staking/availability/measure.go]: what was the cycle? (interfaces also great way to break cycles)

File: staking/availability/measure.go (hunk 2 above)
Comment (rlan35) [reply staking/availability/measure.go]: to avoid wiring unnecessary things into the protocol code.

File: shard/committee/assignment.go [hunk 4]
```diff
@@ -160,6 +168,28 @@ func prepareOrders(
 	return essentials, nil
 }
 
+// IsEligibleForEPoSAuction ..
+func IsEligibleForEPoSAuction(snapshot, validator *staking.ValidatorWrapper) bool {
+	if snapshot.Counters.NumBlocksToSign.Cmp(validator.Counters.NumBlocksToSign) != 0 {
```
Comment (fxfactorial) [inline shard/committee/assignment.go]: so we assume that snapshot of wrapper should have happened already by now, correct? That's only way these will match

File: hmy/api_backend.go [hunk 5]
```diff
@@ -356,13 +356,10 @@ func (b *APIBackend) GetValidatorInformation(
...
-	computed, err := availability.ComputeCurrentSigning(
-		snapshot, wrapper, shard.Schedule.BlocksPerEpoch(),
+	computed := availability.ComputeCurrentSigning(
+		snapshot, wrapper,
 	)
-
-	if err != nil {
-		return defaultReply, nil
-	}
+	computed.BlocksLeftInEpoch = shard.Schedule.BlocksPerEpoch() - computed.ToSign.Uint64()
```
Comment (gupadhyaya) [inline hmy/api_backend.go]: can you also fix this here: https://github.com/harmony-one/harmony/issues/2520

File: staking/availability/measure.go (hunk 1 above)
Comment (rlan35) [reply staking/availability/measure.go]: availability to engine, and engine to availability.

File: shard/committee/assignment.go (hunk 4 above)
Comment (rlan35) [reply shard/committee/assignment.go]: yea, snapshot for elected validator is gauranteed to exist.

File: staking/availability/measure.go (hunk 1 above)
Comment (fxfactorial) [reply staking/availability/measure.go]: Ya this cycle hit me so many times too. So annoying

**Pull request merged**